src/etc/build_rust.py

```diff
@@ -0,0 +1,410 @@
+    rustdoc_path = os.path.join(bin_dir, "rustdoc")
+    aux_base = os.path.join(rust_root, "src", "test", "auxiliary")
+    stage_id = "stage2-" + target
+    test_helpers_lib_dir = glob.glob(
+        os.path.join(target_dir, "debug", "build", "compiletest*", "out"))
+    rust_flags = "--cfg rtopt -O -L " + test_helpers_lib_dir[0]
+    tests_to_run = [("run-pass", "run-pass"),
+                    ("run-pass", "run-pass-fulldeps"),
+                    ("run-pass-valgrind", "run-pass-valgrind"),
+                    ("run-fail", "run-fail"),
+                    ("run-fail", "run-fail-fulldeps"),
+                    ("compile-fail", "compile-fail"),
+                    ("compile-fail", "compile-fail-fulldeps"),
+                    ("parse-fail", "parse-fail"),
+                    ("pretty", "pretty")]
+    gdb_ver = gdb_version()
+    lldb_ver = lldb_version()
+    if gdb_ver:
+        tests_to_run.append(("debuginfo-gdb", "debuginfo"))
+    if lldb_ver:
+        tests_to_run.append(("debuginfo-lldb", "debuginfo"))
+    args = [ctest_exe, "--compile-lib-path", lib_dir,
+            "--run-lib-path", rustlib_dir, "--rustc-path", rustc_path,
+            "--rustdoc-path", rustdoc_path, "--llvm-bin-path", llvm_bin_dir,
+            "--aux-base", aux_base, "--stage-id", stage_id,
+            "--target", target, "--host", host, "--python", sys.executable,
+            "--gdb-version", gdb_ver, "--lldb-version=", lldb_ver,
+            "--android-cross-path", "", "--adb-path", "", "--adb-test-dir", "",
+            "--host-rustcflags", rust_flags, "--target-rustcflags", rust_flags,
+            "--lldb-python-dir", ""]
+    if verbose:
+        args.append("--verbose")
+    test_logs = os.path.join(rust_root, "target", "test-report")
+    shutil.rmtree(test_logs, ignore_errors = True)
+    os.makedirs(test_logs)
+    for test in tests_to_run:
+        src_base = os.path.join(rust_root, "src", "test", test[1])
+        build_base = os.path.join(rust_root, "target", "test", test[1])
+        shutil.rmtree(build_base, ignore_errors = True)
+        os.makedirs(build_base)
+        log_file = os.path.join(test_logs, test[1])
+        final_args = args + ["--src-base", src_base,
+                             "--build-base", build_base,
+                             "--mode", test[0], "--logfile", log_file]
+        ret = subprocess.call(final_args)
+        if ret != 0:
+            print("Compiler test " + test[1] + " failed.")
+
+def build_stage1_rust(rust_root, build,
+                      external_llvm_root, is_release, verbose):
+    print("Building stage1 compiler:")
+    set_env_vars(rust_root, build, external_llvm_root)
+    build_driver_crate(rust_root, build, is_release, verbose)
+    print("Copying stage1 compiler to target/stage1:")
+    build_dir = os.path.join(rust_root, "target")
+    dest_dir = os.path.join(build_dir, "stage1")
+    copy_rust_dist(build_dir, dest_dir, build, [build], is_release)
+    print("Done.")
```

Comment:
Can this function be parameterized over the stag? I believe we'll need to pass `--cfg stageN` to the compiler to get libs to build right in the future, and it may help cut down on duplication.